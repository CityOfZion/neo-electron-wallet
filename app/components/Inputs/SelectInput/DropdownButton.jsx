// @flow
import React from 'react'
import classNames from 'classnames'

import DropdownIcon from '../../../assets/icons/dropdown.svg'
import styles from './DropdownButton.scss'

type Props = {
  className?: string,
  onToggle: Function
}

const DropdownButton = ({ className, onToggle }: Props) => (
  <div className={classNames(styles.dropdownButton, className)} onClick={() => onToggle()}>
    <DropdownIcon className={styles.icon} />
  </div>
)

export default DropdownButton
