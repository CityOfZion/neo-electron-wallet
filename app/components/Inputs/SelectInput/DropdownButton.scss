.dropdownButton {
  display: flex;
  align-items: center;
  height: 100%;
  padding: 0 10px;
  cursor: pointer;

  .icon {
    &, path {
      color: #9599a2;
      fill: #9599a2;
    }
  }
}
