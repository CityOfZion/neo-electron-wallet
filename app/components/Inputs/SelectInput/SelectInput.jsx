// @flow
import React from 'react'
import classNames from 'classnames'
import { noop, omit, trim } from 'lodash'
import Sifter from 'sifter'

import Dropdown from './Dropdown'
import styles from './SelectInput.scss'

type Props = {
  className?: string,
  value?: string,
  items?: Array<Object>,
  renderItem?: Function,
  renderAfter?: Function,
  onFocus?: Function,
  onChange?: Function
}

type State = {
  open: boolean,
  searchTerm: string
}

export default class SelectInput extends React.Component<Props, State> {
  static defaultProps = {
    items: [],
    onFocus: noop,
    onChange: noop
  }

  state = {
    open: false,
    search: this.props.value || ''
  }

  componentWillReceiveProps = (nextProps) => {
    if (nextProps.value !== this.props.value) {
      this.setState({ search: trim(nextProps.value) })
    }
  }

  render = () => {
    const passDownProps = omit(this.props, 'items', 'renderItem', 'renderAfter')

    return (
      <Dropdown
        className={classNames(styles.selectInput, this.props.className)}
        open={this.state.open}
        onClose={this.handleClose}
        renderDropdown={this.renderDropdown}>
        <div className={styles.inputContainer}>
          <input
            {...passDownProps}
            className={styles.input}
            type='text'
            onFocus={this.handleFocus}
            onChange={this.handleChange}
          />
          {this.renderAfter()}
        </div>
      </Dropdown>
    )
  }

  renderDropdown = ({ className }) => {
    const items = this.getItems()
    const hasItems = items.length > 0
    const isSearch = this.state.search.length > 0 && this.props.items.length > 0

    if (hasItems) {
      return (
        <div className={classNames(styles.dropdown, className)}>
          {this.renderItems(items)}
        </div>
      )
    } else if (isSearch) {
      return (
        <div className={classNames(styles.dropdown, styles.noSearchResults, className)}>
          No search results.
        </div>
      )
    } else {
      return null
    }
  }

  renderItems = (items) => {
    return items.map((item) => {
      const renderItem = this.props.renderItem || this.renderItem
      return renderItem(item, { onSelect: this.generateSelectHandler(item) })
    })
  }

  renderItem = (item, { onSelect }) => {
    return (
      <div className={styles.dropdownItem} key={item.value} tabIndex={0} onClick={onSelect}>
        <div className={styles.label}>{item.label}</div>
        <div className={styles.value}>{item.value}</div>
      </div>
    )
  }

  renderAfter = () => {
    if (this.props.renderAfter) {
      return (
        <div className={styles.afterInput}>
          {this.props.renderAfter()}
        </div>
      )
    }
  }

  handleFocus = (event: Event, ...args: Array<any>) => {
    this.props.onFocus(event, ...args)
    this.setState({ open: true })
  }

  handleChange = (event: Event, ...args: Array<any>) => {
    this.props.onChange(event.target.value)
  }

  handleClose = () => {
    this.setState({ open: false })
  }

  generateSelectHandler = (item) => {
    return (event) => {
      this.props.onChange(item.value)
      this.setState({ open: false })
    }
  }

  getItems = () => {
    const { items } = this.props
    const sifter = new Sifter(items)
    const result = sifter.search(this.state.search, {
      fields: ['label', 'value'],
      sort: [{ field: 'label', direction: 'asc' }]
    })
    return result.items.map((resultItem) => items[resultItem.id])
  }
}
