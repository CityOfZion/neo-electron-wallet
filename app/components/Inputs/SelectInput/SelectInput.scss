@import '../../../styles/variables';

.selectInput {
  position: relative;

  &.error {
    border: 2px solid $error-color;
    border-top-right-radius: 3px;
    border-top-left-radius: 3px;
  }

    .errorIcon {
    color: #EE6D66;
    position: absolute;
    right: 8px;
    top: 10px;
  }

  .errorMessage {
    background-color: $error-color;
    color: #fff;
    position: absolute;
    top: 39px;
    left: -2px;
    width: 100%;
    font-size: 0.7rem;
    padding: 5px 2px;
    border-radius: 3px;
    z-index: 9999;
    text-align: center;
  }

  .dropdown {
    box-sizing: border-box;
    width: 100%;
    max-height: 180px;
    overflow: auto;
    background: #f2f2f2;
    border: 2px solid #323242;
    border-top: none;
    border-radius: 0 0 4px 4px;
    z-index: 1;

    .dropdownItem {
      padding: 10px 20px;
      color: #000;
      font-size: 14px;
      cursor: pointer;

      &:hover {
        background: #e6e6e6;
      }

      .highlight {
        font-weight: 600;
      }
    }

    &.noSearchResults {
      padding: 10px 20px;
      color: #9e9e9e;
      text-align: center;
      font-size: 14px;
      font-style: italic;
    }
  }
}
