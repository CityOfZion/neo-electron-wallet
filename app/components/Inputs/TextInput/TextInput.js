// @flow
import React from 'react'
import classNames from 'classnames'
import { omit } from 'lodash'

import styles from './TextInput.scss'

type Props = {
  className: ?string,
  type: string,
  renderBefore: ?Function,
  renderAfter: ?Function
}

export default class TextInput extends React.Component<Props> {
  static defaultProps = {
    type: 'text'
  }

  render () {
    const passDownProps = omit(this.props, 'className', 'renderBefore', 'renderAfter')

    return (
      <div className={classNames(styles.textInput, this.props.className)}>
        {this.renderBefore()}
        <input {...passDownProps} className={styles.input} />
        {this.renderAfter()}
      </div>
    )
  }

  renderBefore = () => {
    if (!this.props.renderBefore) {
      return null
    }

    return (
      <div className={styles.beforeInput}>
        {this.props.renderBefore()}
      </div>
    )
  }

  renderAfter = () => {
    if (!this.props.renderAfter) {
      return null
    }

    return (
      <div className={styles.afterInput}>
        {this.props.renderAfter()}
      </div>
    )
  }
}
