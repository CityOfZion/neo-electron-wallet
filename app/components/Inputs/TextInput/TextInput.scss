@import '../../../styles/variables';

.textInput {
  $text-input-color: #282828;
  $text-input-background-color: #f2f2f2;
  $text-input-border-radius: 3px;
  $text-input-height: 36px;
  $text-input-font-size: 12px;

  display: flex;
  flex-wrap: nowrap;
  align-items: stretch;
  height: $text-input-height;
  border-radius: $text-input-border-radius;
  background: $text-input-background-color;
  overflow: hidden;

  .beforeInput {
    flex: 0 1 auto;
  }

  .input {
    flex: 1 1 auto;
    box-sizing: border-box;
    width: 100%;
    height: $text-input-height;
    line-height: $text-input-height;
    color: $text-input-color;
    background: $text-input-background-color;
    font-size: $text-input-font-size;
  }

  .afterInput {
    flex: 0 1 auto;
  }
}
