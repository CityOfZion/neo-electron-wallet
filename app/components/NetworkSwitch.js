import React, { Component } from 'react';
import { connect } from 'react-redux';
import { setNetwork } from '../actions/index.js';
import { getBalance, getTransactionHistory, getMarketPriceUSD, neoId, getClaimAmounts, getWalletDBHeight } from '../wallet/api.js';
import { setBalance, setMarketPrice, resetPrice, setTransactionHistory, setClaim, setBlockHeight } from '../actions/index.js';

let intervals = {};

let netSelect;

// TODO: this is being imported by Balance.js, maybe refactor to helper file

const initiateGetBalance = (dispatch, net, address) => {
  syncTransactionHistory(dispatch, net, address);
  syncAvailableClaim(dispatch, net, address);
  syncBlockHeight(dispatch, net);
  return getBalance(net, address).then((resultBalance) => {
    return getMarketPriceUSD(resultBalance.Neo).then((resultPrice) => {
      dispatch(setBalance(resultBalance.Neo, resultBalance.Gas, resultPrice));
      return true;
    });
  }).catch((result) => {
    // TODO: is this ever called?
  });
};

const syncAvailableClaim = (dispatch, net, address) => {
  getClaimAmounts(net, address).then((result) => {
    //claimAmount / 100000000
    dispatch(setClaim(result.available, result.unavailable));
  });
}

const syncBlockHeight = (dispatch, net) => {
  getWalletDBHeight(net).then((blockHeight) => {
    dispatch(setBlockHeight(blockHeight));
  });
};

const syncTransactionHistory = (dispatch, net, address) => {
  getTransactionHistory(net, address).then((transactions) => {
    let txs = [];
    for (let i = 0; i < transactions.length; i++){
      if (transactions[i].neo_sent === true){
        txs = txs.concat([{type: "NEO", amount: transactions[i].NEO, txid: transactions[i].txid, block_index: transactions[i].block_index }]);
      }
      if (transactions[i].gas_sent === true){
        txs = txs.concat([{type: "GAS", amount: transactions[i].GAS, txid: transactions[i].txid, block_index: transactions[i].block_index }]);
      }
    }
    dispatch(setTransactionHistory(txs));
  });
};

const resetBalanceSync = (dispatch, net, address) => {
  console.log("sync", net);
  if (intervals.balance !== undefined){
    clearInterval(intervals.balance);
  }
  intervals.balance = setInterval(() =>  {
    initiateGetBalance(dispatch, net, address);
  }, 5000);
};

const toggleNet = (dispatch, net, address) => {
  let newNet;
  if (net === "MainNet"){
    newNet = "TestNet";
  } else {
    newNet = "MainNet";
  }
  dispatch(setNetwork(newNet));
  resetBalanceSync(dispatch, newNet, address);
  if (address !== null){
    initiateGetBalance(dispatch, newNet, address);
  }
};

class NetworkSwitch extends Component {
  componentDidMount = () => {
    resetBalanceSync(this.props.dispatch, this.props.net, this.props.address);
  }

  render = () =>
    <div id="network">
      <span className="transparent">Running on</span>
      <span className="netName" onClick={() => toggleNet(this.props.dispatch, this.props.net, this.props.address)}>{this.props.net}</span>
    </div>;

}

const mapStateToProps = (state) => ({
  net: state.metadata.network,
  address: state.account.address
});

NetworkSwitch = connect(mapStateToProps)(NetworkSwitch);

export { NetworkSwitch, initiateGetBalance, syncTransactionHistory, intervals };
