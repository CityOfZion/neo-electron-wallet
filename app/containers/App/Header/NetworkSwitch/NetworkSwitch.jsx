// @flow
import React, { Component } from 'react'

import headerStyles from '../Header.scss'

type Props = {
  networkId: string,
  onChange: Function,
  networks: Array<NetworkItemType>
}

export default class NetworkSwitch extends Component<Props> {
  render () {
    const { networkId, networks } = this.props
    return (
      <div id='network' className={headerStyles.navBarItem}>
        <span className={headerStyles.navBarItemLabel}>Running on</span>
        <select defaultValue={networkId} onChange={this.handleChange} className='networkSelector'>
          {networks.map(({ label, id }: NetworkItemType) =>
            <option key={`networkOption${id}`} value={id}>{label}</option>
          )}
        </select>
      </div>
    )
  }

  handleChange = (event: Object) => {
    this.props.onChange(event.target.value)
  }
}
