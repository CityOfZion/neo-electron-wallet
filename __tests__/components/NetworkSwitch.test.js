import React from 'react'
import { mount } from 'enzyme'

import NetworkSwitch from '../../app/containers/App/Sidebar/NetworkSwitch/NetworkSwitch'
import TextInput from '../../app/components/Inputs/TextInput'
import SelectInput from '../../app/components/Inputs/SelectInput'
import DropdownButton from '../../app/components/Inputs/SelectInput/DropdownButton'
import { MAIN_NETWORK_ID, TEST_NETWORK_ID } from '../../app/core/constants'

const setup = () => {
  const props = {
    networkId: MAIN_NETWORK_ID,
    networks: [
      {
        id: MAIN_NETWORK_ID,
        label: 'MainNet',
        network: 'MainNet'
      },
      {
        id: TEST_NETWORK_ID,
        label: 'TestNet',
        network: 'TestNet'
      }
    ],
    onChange: jest.fn()
  }
  const wrapper = mount(<NetworkSwitch {...props} />)

  return {
    wrapper
  }
}

describe('NetworkSwitch', () => {
  test('renders without crashing', () => {
    const { wrapper } = setup()
    expect(wrapper).toMatchSnapshot()
  })

  test('correctly renders MainNet initially', () => {
    const { wrapper } = setup()
    const networkInput = wrapper.find(TextInput).instance()

    expect(networkInput.props.value).toEqual('MainNet')
  })

  test('switches to the correct network when chosen from the dropdown', () => {
    const { wrapper } = setup()
    const networkSelector = wrapper.find(SelectInput).instance()

    networkSelector.setState({ open: true })

    wrapper.update()
    const dropDownButton = wrapper.find('.dropdownItem')
    dropDownButton.simulate('click')

    expect(wrapper.instance().props.onChange).toHaveBeenCalledWith(
      TEST_NETWORK_ID
    )
  })
})
