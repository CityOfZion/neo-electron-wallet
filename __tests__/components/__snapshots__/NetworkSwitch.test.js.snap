// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NetworkSwitch renders without crashing 1`] = `
<NetworkSwitch
  networkId="1"
  networks={
    Array [
      Object {
        "id": "1",
        "label": "MainNet",
        "network": "MainNet",
      },
      Object {
        "id": "2",
        "label": "TestNet",
        "network": "TestNet",
      },
    ]
  }
  onChange={[MockFunction]}
>
  <div
    className="networkSwitch"
    id="network"
  >
    <SelectInput
      customChangeEvent={true}
      getItemValue={[Function]}
      getSearchResults={[Function]}
      items={
        Array [
          Object {
            "id": "2",
            "label": "TestNet",
            "network": "TestNet",
          },
        ]
      }
      onChange={[Function]}
      onFocus={[Function]}
      readOnly={true}
      renderAfter={[Function]}
      renderItem={[Function]}
      value="MainNet"
    >
      <clickOutside(Dropdown)
        className="selectInput"
        onClose={[Function]}
        open={false}
        renderDropdown={[Function]}
      >
        <Dropdown
          className="selectInput"
          onClose={[Function]}
          open={false}
          renderDropdown={[Function]}
        >
          <div
            className="dropdown selectInput"
          >
            <div
              className="anchor"
            >
              <TextInput
                className="input "
                onChange={[Function]}
                onFocus={[Function]}
                readOnly={true}
                renderAfter={[Function]}
                type="text"
                value="MainNet"
              >
                <div
                  className="textInputContainer"
                >
                  <div
                    className="textInput input "
                  >
                    <input
                      className="input "
                      onBlur={[Function]}
                      onChange={[Function]}
                      onFocus={[Function]}
                      readOnly={true}
                      type="text"
                      value="MainNet"
                    />
                    <div
                      className="afterInput"
                    >
                      <DropdownButton
                        onToggle={[Function]}
                      >
                        <Component
                          className="dropdownButton"
                          onClick={[Function]}
                        >
                          <svg
                            className="dropdownButton"
                            onClick={[Function]}
                          />
                        </Component>
                      </DropdownButton>
                    </div>
                  </div>
                </div>
              </TextInput>
            </div>
          </div>
        </Dropdown>
      </clickOutside(Dropdown)>
    </SelectInput>
  </div>
</NetworkSwitch>
`;
